### PR TITLE
Improved performance for multiple negation filters

### PR DESCRIPTION
**Pull request by 9larsons**
ref https://linear.app/tryghost/issue/CFR-27
- NQL will now automatically combine multiple not equal ($ne) filters of the same type
- added JSDoc

This is a performance improvement as (NOT 1) AND (NOT 2) is logically equal to NOT (1 OR 2), except the latter is *much* more performant in the resulting SQL query as you only get one subquery instaed of one for each filter.

### CODE REVIEW

**Comment by 9larsons** [comment]
[This](https://github.com/TryGhost/NQL/pull/78/files#diff-a4c6d751906029602b36abb79c51a2388ae8055abcc52220f0de95b904032249R48-R68) set of lines is the meat of the changes.

File: packages/nql/lib/nql.js [hunk 1]
```diff
@@ -31,32 +49,52 @@ module.exports = (queryString, options = {}) => {
 
         let mongoJSON = utils.mergeFilters(overrides, this.filter, defaults);
 
+        // this is a performance modification to combine multiple $ne filters into a single $nin filter (see de Morgan's laws)
```
Comment (cmraible) [inline packages/nql/lib/nql.js:52]: > (see de Morgan's laws)

👨‍🍳💋

**Review by cmraible — APPROVED** [review]
@9larsons LGTM! Thanks for adding all the JSDoc comments, they provided a helpful crash course in NQL 😃

I've been trying to break this for about an hour and came up empty — nice work!

**Review by ErisDS — APPROVED** [review]
First, lemme say it makes me SO HAPPY to see someone diving into this codebase, and to have a chance to fiddle with it myself 😄

You've nailed all of the changes here 👌 

I absolute love that you have taken the time to "Scout rule" and improve the comments to make this more accessible to the next person. 

As a minor detail, it's my preference to do scout rule commits separately and push them straight to main and keep the main change set clean _however_ this is perhaps something to bring up as a Working Agreement and see if I am alone on this!

I had a few questions about the logic as I read through this, and the easiest way for me to answer them was to write a few extra tests, which I've pushed up to the PR in the interest of speed.

Anyway, TLDR this change is perfect & really exciting. I think there's so many more of these the CFR team can add and our aborted get helpers will start to disappear 🙌
